275 pgsqlsearch

leonstr opened the pull request:
PR with proposed fix for #275 

This is built on the fix for #273 as the CI check failures for that PR (#274 ) are actually tool_mergeusers bugs, the ones fixed in this commit.

leonstr committed:
273 Excluded deleted users from search results

leonstr committed:
275 Fix invalid syntax error in PostgreSQL

Previously searching for users with "Id" or "All" selected but non-
numeric criteria resulted in "ERROR: invalid input syntax for type
bigint" when using PostgreSQL.

jpahullo commented on lib/mergeusersearch.php:
```diff
@@ -60,10 +60,16 @@ public function search_users($input, $searchfield){
-                $params = array(
-                    'userid' => $input,
-                );
-                $sql = 'SELECT * FROM {user} WHERE id = :userid';
```
Thanks @leonstr for this patch. But, unfortunatelly, this is not the way of doing. By intheritance of the beginning of the development of this plugin, there are some places with specific code and conditions according to types of database. 

However, the moodle_database class affords this specificities very well. What we need to call, however, are the proper API to the `$DB` instance, like the `$DB->get_in_or_equal(...)` that already catches these particularities of null values and other specific dababase API.

I mean, the error you are providing the patch is that we originally did not considered using the `$DB` API properly. 

Could you, please, use the `$DB->get_in_or_equal()` API and remake this part?

jpahullo replied:
In addition, this DB API would also consider edge cases, like empty strings, null values and so, producing the necessary comparison condition for the WHERE clause. Currently the ctype_digit() call matches both for empty string and null values. Empty string would be a false positive.

leonstr replied:
I investigated `$DB->get_in_or_equal()` but I don't think this handles the issue matching strings with integer columns in PostgreSQL. In fact, this only does any special handling for this database type if there are >= 65535 elements in the `IN (...)`.

I have reworked the use of `ctype_digit()` to `is_string($input) && !ctype_digit($input)` which should now correctly handle, e.g., "", "123", 123, and null.

I've also reworked `MergeUserSearch->search_users()` to use `$DB->get_records_select()` instead of `$DB->get_records_sql().

jpahullo commented on lib/mergeusersearch.php:
```diff
@@ -107,9 +113,15 @@ public function search_users($input, $searchfield){
 
                 break;
             default: // search on all fields by default
+                if (!ctype_digit($input)) {
```
Idem as before. Use `$DB->get_in_or_equal()` API to build this part of the SQL and the corresponding array of parameters.

leonstr replied:
Again, I don't think `$DB->get_in_or_equal()` helps us here, I've reworked the use of `ctype_digit()` to `is_string($input) && !ctype_digit($input)`.

leonstr replied:
Thinking about this again my original approach could be improved by using `$DB->sql_cast_to_char()` so instead of `WHERE id = 'abc'` which gets an error in PostgreSQL, we now have `WHERE CAST(id AS VARCHAR) = 'abc'` which avoids this error. So I've added a further commit (09994f2) with this change.

(All CI tests have failed but I don't think that's to do with my commit).

jpahullo commented on lib/mergeusersearch.php:
```diff
@@ -118,19 +130,20 @@ public function search_users($input, $searchfield){
                     FROM {user}
                     WHERE
-                        id = :userid OR
+                        ($whereid OR
```
Revisit according to the above changes, if necessary.

leonstr replied:
This is now `'id = :userid OR` again.

jpahullo commented on tests/search_users_test.php:
```diff
@@ -0,0 +1,130 @@
+        global $DB;
+
+        // Skip tests if not using PostgreSQL.
+        if ($DB->get_dbfamily() != 'postgres') {
```
Tests should be generic and database agnostic, and not database specific. We should remove this condition. The code should work on any supported database family. If we apply the above changes, this test automatically will be generic.

leonstr replied:
I've removed the PostgreSQL-specific test, this was a poorly-thought-out addition on my part.

jpahullo replied:
thanks!

jpahullo commented on tests/search_users_test.php:
```diff
@@ -0,0 +1,130 @@
+     * error when comparing and integer column (mdl_user.id) with a string.
+     * @dataProvider search_criteria
+     */
+    public function test_pgsqlsearch($searchfield, $input, $count): void {
```
rename the test name, please, as you had before in the other PR.

leonstr replied:
There's now only the `test_nodeletedusers` test.

jpahullo commented on tests/search_users_test.php:
```diff
@@ -0,0 +1,130 @@
+                'input' => '',
+                'count' => 0,
+            ],
+            'id2' => [
```
I can see that you removed the test with null as input value. Even though it would be an edge case (not possible using Moodle administration to merge users), it can be possible if someone integrates it programatically and undetectedly pass some null value over there.  Please, add again that test case with null as input value. Also, removing this test case, your patch to distinct between id IS NULL and so, does not apply now. Actually, it would apply to empty string (since it is not a number), but is a false positive.

leonstr replied:
I've increased the data in the data provider, it includes `'input' => null`.

jpahullo replied:
thanks!

jpahullo reviewed (changes requested):
Please, consider my comments to adapt the patch before it could get merged.

I liked your initial idea for the postgresql specificities, but DB API from Moodle provides already the proper solution. Read my comments for more details.

Thanks a lot for your work.

leonstr committed:
Rework search_users() to use get_records_select()

Rework MergeUserSearch->search_users() to better use the Moodle DML API.

Also increased the range of data used in the PHPUnit test for the search
users functionality.

jpahullo commented:
Thanks @leonstr for contributing this patch.

I need to test it in our testing Moodle instance carefully. And if applicable, release a new plugin version.

Unfortunately, we are going to vacation. Then, until September probably it will not land into the plugin in any case. 

I will let you know of its progress.

Thanks a lot for your hard work.

Jordi

leonstr committed:
Rework search_users() to use sql_cast_to_char()

Rework MergeUserSearch->search_users() to better handing comparing
mdl_user.id when the user-supplied value is not an integer.
